File: docs/mcoimg-js/tests/verify-v3-js-encoder-fixtures.js
'use strict';

const fs = require('node:fs');
const path = require('node:path');
const os = require('node:os');
const { spawnSync } = require('node:child_process');

const caseArgument = process.argv.find((argument) => argument.startsWith('--case='));
const caseIndex = caseArgument == null ? null : Number(caseArgument.slice('--case='.length));
const positionalArguments = process.argv.slice(2).filter((argument) => !argument.startsWith('--'));
const fixturePath = positionalArguments[0]
  ? path.resolve(positionalArguments[0])
  : path.join(__dirname, 'v3-js-encoder-fixtures.json');
const document = JSON.parse(fs.readFileSync(fixturePath, 'utf8'));

if (caseIndex == null) {
  const summaries = [];
  for (let index = 0; index < document.fixtures.length; index++) {
    const tempDirectory = fs.mkdtempSync(path.join(os.tmpdir(), 'mcoimg-v3-verify-'));
    const stdoutPath = path.join(tempDirectory, 'stdout.json');
    const stderrPath = path.join(tempDirectory, 'stderr.txt');
    const stdoutFd = fs.openSync(stdoutPath, 'w');
    const stderrFd = fs.openSync(stderrPath, 'w');
    let child;
    try {
      child = spawnSync(process.execPath, [__filename, fixturePath, `--case=${index}`], {
        stdio: ['ignore', stdoutFd, stderrFd],
        timeout: 180000,
      });
    } finally {
      fs.closeSync(stdoutFd);
      fs.closeSync(stderrFd);
    }
    const childStdout = fs.readFileSync(stdoutPath, 'utf8');
    const childStderr = fs.readFileSync(stderrPath, 'utf8');
    fs.rmSync(tempDirectory, { recursive: true, force: true });
    if (child.error) throw child.error;
    if (child.status !== 0) {
      throw new Error(`fixture ${index} failed:
${childStderr || childStdout}`);
    }
    summaries.push(JSON.parse(childStdout));
  }
  console.log(JSON.stringify({
    fixtureFile: path.basename(fixturePath),
    generator: document.generator,
    fixtures: summaries,
    exactReencode: 'ok',
  }, null, 2));
  process.exit(0);
}

if (!Number.isInteger(caseIndex) || caseIndex < 0 || caseIndex >= document.fixtures.length) {
  throw new RangeError(`invalid --case value: ${caseArgument}`);
}

require(path.resolve(__dirname, '..', 'mcoimg-v3-codec.global.js'));
const v3 = globalThis.MCOImgV3;
const codec = new v3.MCOImageV3Codec();

function same(left, right) {
  return left.length === right.length && left.every((value, index) => value === right[index]);
}

function verifyFixture(index) {
  const fixture = document.fixtures[index];
  const profile = v3.PaletteProfile[fixture.paletteProfile];
  const expectedBody = new Uint8Array(Buffer.from(fixture.bodyBase64, 'base64'));
  const expectedApp = new Uint8Array(Buffer.from(fixture.appPayloadBase64, 'base64'));
  const encoded = codec.encode({
    width: fixture.width,
    height: fixture.height,
    paletteProfile: profile,
    pixels: fixture.pixels,
    transparentColor: fixture.transparentColor ?? null,
  }, {
    compressionLevel: fixture.compressionLevel,
    packetNonce: fixture.packetNonce,
  });
  if (!same(encoded.body, expectedBody)) {
    throw new Error(`${fixture.name}: checked-in body no longer re-encodes exactly`);
  }
  if (!same(encoded.appPayloadWithoutSender, expectedApp)) {
    throw new Error(`${fixture.name}: checked-in app payload mismatch`);
  }
  if (encoded.text !== fixture.text) {
    throw new Error(`${fixture.name}: checked-in text mismatch`);
  }
  const decoded = codec.decodeBody(expectedBody);
  if (decoded.width !== fixture.width || decoded.height !== fixture.height ||
      decoded.paletteProfile !== profile ||
      decoded.transparentColor !== (fixture.transparentColor ?? null) ||
      !same(decoded.pixels, fixture.pixels)) {
    throw new Error(`${fixture.name}: checked-in fixture decode mismatch`);
  }
  const info = v3.MCOImageV3Codec.inspectBody(expectedBody);
  return {
    name: fixture.name,
    level: fixture.compressionLevel,
    bytes: expectedBody.length,
    container: info.containerName,
    algorithm: info.blockAlgorithmName,
  };
}

process.stdout.write(`${JSON.stringify(verifyFixture(caseIndex))}\n`, () => process.exit(0));

File: docs/mcoimg-js/tests/verify_v3_js_encoder_fixtures.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:meshcore_open/helpers/mcoimg_types.dart';
import 'package:meshcore_open/helpers/mcoimg_v3_codec.dart';

PaletteProfile _paletteProfile(String name) {
  return PaletteProfile.values.firstWhere(
    (value) => value.name == name,
    orElse: () => throw ArgumentError('Unknown palette profile: $name'),
  );
}

Future<void> main(List<String> arguments) async {
  final scriptDirectory = File.fromUri(Platform.script).parent;
  final fixtureFile = File(
    arguments.isNotEmpty
        ? arguments.first
        : '${scriptDirectory.path}/v3-js-encoder-fixtures.json',
  );
  final document = jsonDecode(await fixtureFile.readAsString()) as Map<String, dynamic>;
  final fixtures = document['fixtures'] as List<dynamic>;
  final codec = MCOImageV3Codec();
  final summaries = <Map<String, dynamic>>[];

  for (final raw in fixtures) {
    final fixture = Map<String, dynamic>.from(raw as Map);
    final name = fixture['name'] as String;
    final body = Uint8List.fromList(base64Decode(fixture['bodyBase64'] as String));
    final appPayload = Uint8List.fromList(
      base64Decode(fixture['appPayloadBase64'] as String),
    );
    final profile = _paletteProfile(fixture['paletteProfile'] as String);
    final pixels = List<int>.from(fixture['pixels'] as List);
    final transparentColor = fixture['transparentColor'] as int?;

    final decoded = <String, MCOImage>{
      'body': codec.decodeBody(body),
      'app': codec.decodeAppPayloadWithoutSender(appPayload),
      'text': codec.decodeText(fixture['text'] as String),
    };
    for (final entry in decoded.entries) {
      final image = entry.value;
      if (image.width != fixture['width'] ||
          image.height != fixture['height'] ||
          image.paletteProfile != profile ||
          image.transparentColor != transparentColor ||
          !_sameInts(image.pixels, pixels)) {
        throw StateError('$name: Dart ${entry.key} decode mismatch');
      }
    }
    if (!_sameInts(MCOImageV3Codec.bodyFromText(fixture['text'] as String), body)) {
      throw StateError('$name: Dart text/body conversion mismatch');
    }
    summaries.add(<String, dynamic>{
      'name': name,
      'bytes': body.length,
      'container': fixture['container'],
      'algorithm': fixture['algorithm'],
    });
  }

  stdout.writeln(const JsonEncoder.withIndent('  ').convert(<String, dynamic>{
    'fixtureFile': fixtureFile.path,
    'generator': document['generator'] ?? 'unknown',
    'fixtures': summaries,
  }));
}

bool _sameInts(List<int> left, List<int> right) {
  if (left.length != right.length) return false;
  for (var index = 0; index < left.length; index++) {
    if (left[index] != right[index]) return false;
  }
  return true;
}
